File: Лаба 5/2.py
fin = open("check.in")
fout = open("check.out", "w")

def check(i, k, flag):
    F = True
    if L[i] >= 0:
        F1 = (K[L[i]] < k) if flag else (K[L[i]] > k)
        F = F1 and (K[i] > K[L[i]]) and check(L[i], k, flag)
    if R[i] >= 0 and F:
        F1 = (K[R[i]] < k) if flag else (K[R[i]] > k)
        F = F1 and (K[i] < K[R[i]]) and check(R[i], k, flag)
    return F

n = int(fin.readline())

K = []
L = []
R = []

for i in range(n):
    k, l, r = map(int, fin.readline().split())
    K.append(k)
    L.append(l - 1)
    R.append(r - 1)

correct = True
if n > 0:
    top = 0

    if L[top] >= 0:
        if K[L[top]] < K[top]:
            correct = check(L[top], K[top], True)
        else:
            correct = False

    if correct and R[top] >= 0:
        if K[R[top]] > K[top]:
            correct = check(R[top], K[top], False)
        else:
            correct = False

if correct:
    print('YES', file=fout)
else:
    print('NO', file=fout)

fin.close()
fout.close()

File: Лаба 5/2.cpp
#include <fstream>
#include <vector>
using namespace std;

struct Tree {
	long long k;
	int l;
	int r;
};

bool Flag = true;
vector<Tree> my_tree;

void check(int v, long long min, long long max) {
	if (my_tree[v].k <= min || my_tree[v].k >= max) {
		Flag = false;
	}

	if (my_tree[v].l != 0) {
		check(my_tree[v].l, min, my_tree[v].k);
	}

	if (my_tree[v].r != 0) {
		check(my_tree[v].r, my_tree[v].k, max);
	}
}

int main() {
	ifstream fin("check.in");
	ofstream fout("check.out");

	int n;
	fin >> n;

	if (n == 0){
		fout << "YES";
	}
	else {
		my_tree.resize(n + 1);/*чтобы нумерация шла от 1 до n, 
		а не от 0 до n-1*/

		for (int i = 1; i <= n; i++) {
			fin >> my_tree[i].k >> my_tree[i].l >> my_tree[i].r;
		}

		check(1, -1000000001, 1000000001);

		if (Flag) {
			fout << "YES";
		}
		else {
			fout << "NO";
		}
	}
}